packages/transport-tcp/src/listener.ts

```diff
@@ -304,20 +306,28 @@ export class TCPListener extends TypedEventEmitter<ListenerEvents> implements Li
   }
 
   async close (): Promise<void> {
-    const err = new AbortError('Listener is closing')
+    const events: Array<Promise<void>> = []
 
-    // synchronously close each connection
-    this.connections.forEach(conn => {
-      conn.abort(err)
-    })
+    if (this.server.listening) {
+      events.push(pEvent(this.server, 'close'))
+    }
+
+    // shut down the server socket, permanently
+    this.pause(true)
+
+    // stop any in-progress connection upgrades
+    this.shutdownController.abort()
 
-    // cleanup connections that have not been upgraded
-    this.maConnections.forEach(conn => {
-      conn.abort(err)
+    // synchronously close any open connections - should be done after closing
+    // the server socket in case new sockets are opened during the shutdown
+    this.sockets.forEach(socket => {
+      if (socket.readable) {
+        events.push(pEvent(socket, 'close'))
+        socket.destroy()
+      }
     })
```

Comment:
If the shutdownController is aborted, would this cause the catch block to be called on the upgrade in onSocket? That block removes the socket from this.sockets. Wondering if this could be an issue?

Reply:
Ah i see the maConn is aborted directly after.